core/src/main/scala/cr/pulsar/Subscription.scala

```diff
@@ -71,11 +73,25 @@ object Subscription {
     }
   }
 
-  def apply(name: Name, sType: Type, mode: Mode) =
-    new Subscription(
-      name.value ++ "-subscription",
-      sType.pulsarSubscriptionType,
-      mode.pulsarSubscriptionMode
-    ) {}
+  private case class SubscriptionImpl(
+      name: Subscription.Name,
+      `type`: Subscription.Type,
+      mode: Subscription.Mode
```

Comment:
It seems this list starts to grow and potentially it can grow even more. We have `Options` for `Producer` implementation, should we use the same pattern here?

Reply:
mmm this uses the same pattern as `Options`:

```scala
sealed abstract class Subscription {
  val name: Subscription.Name
  val `type`: Subscription.Type
  val mode: Subscription.Mode
  def withType(_type: Subscription.Type): Subscription
  def withMode(_mode: Subscription.Mode): Subscription
}
```

Or you have something else in mind?